lib/Debug/DebugInfo.cpp

```diff
@@ -0,0 +1,34 @@
+/// \file DebugInfo.cpp
+/// \brief Run an instance of DebugHelper over the IR.
+
+#include "llvm/Pass.h"
+#include "llvm/IR/AssemblyAnnotationWriter.h"
+#include "llvm/IR/IRBuilder.h"
+
+#include "revng/Support/CommandLine.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/DebugHelper.h"
+#include "revng/Support/revng.h"
+
+using namespace llvm;
+
+namespace {
+
+class DebugInfo : public ModulePass {
+public:
+    static char ID;
+    DebugInfo() : ModulePass(ID) { }
+
+    bool runOnModule(Module &M);
+};
+
+}
+
+bool DebugInfo::runOnModule(Module &M) {
+    DebugHelper DBG("debug_info", &M, DebugInfoType::LLVMIR, M.getModuleIdentifier());
```

Comment:
* [ ] Why `getModuleIdentifier`?
* [ ] Is this creating a `debug_info` file in the current directory?

This pass should not create any file, it should just decorate the IR, AFAIU.

Reply:
The pass doesn't create any file, and no `debug_info` file is created, it simply decorates the IR. ` M.getModuleIdentifier()` comes because we want to overwrite the current `.ll` with the new annotations. If `-debug-info` is enabled, I don't think it would make sense to create a separate `.ll.with_debug_decorations` file.

Reply:
> M.getModuleIdentifier() comes because we want to overwrite the current .ll with the new annotations

I don't understand. This is a pass, it's not supposed to overwrite `opt` output.

```c++
  /// \brief Create a new DebugHelper
  ///
  /// \param Output path where the LLVM IR should be stored.
  /// \param Debug path where the debug output should be stored. If empty, \p
  ///        Output will be used along with a suffix, e.g. `.pts` if \p Type is
  ///        DebugInfoType::PTC, `.S` if it's DebugInfoType::OriginalAssembly or
  ///        will match \p Output if \p Type is DebugInfoType::LLVMIR.
  /// \param TheModule the LLVM module to print out.
  DebugHelper(std::string Output,
              llvm::Module *TheModule,
              DebugInfoType::Values DebugInfo,
              std::string DebugPath);
```
You're passing "debug_info" as `Output`. Are you sure it's not creating a file?

```c++
    std::ofstream Output(DebugPath);
    raw_os_ostream Stream(Output);
    TheModule->print(Stream, annotator(false));
```

Why are you creating a file named `M.getModuleIdentifier()` (`DebugPath` argument)?

I'm confused.

Reply:
I came to the conclusion that this is not meant to go upstream. First, the user shouldn't likely need to debug the IR. Second, I didn't really bother to tell if the pass should overwrite `opt` output or not. As things stand, the `.ll` happens to be fully overwritten with the addition of the debug information. I agree that the IR should be simply changed with the debug info, without re-creating the entire module but it was not within the goals when I created that pass. Maybe @andrealinux1 has a say in whole of this, since I adapted this pass from some of his old stuff; and by now this dates back to more than a year ago.

Reply:
Maybe the right approach would be:

1. Run the pass adding metadata, no file is written.
2. Have another pass that has an extra option for the output path, where the decorated file is dumped.

Sadly, we can't do much more than this, since to decorate `opt` output we'd need our custom annotator.
The feature is useful and interesting, however, if you prefer to remove this from this PR, I agree. It's already quite big.

Reply:
To be reviewed in a second moment.

Reply:
Mmmh, I'm also a bit confused, so forgive me if I say trivial things or I make stupid annotations.

I guess here the problem is that, as things are now, this new pass creates a new `.ll` file named as the module itself (basically, it overwrites it), due to the fact that we pass explicitly the `DebugPath` parameter. If this is not done, what @aleclearmind says it's right, the pass creates a new `debug_info` file.

What @aleclearmind is suggesting, and tell me if I got this right, is that, following the conventions of llvm passes, a pass should not create a new output per se, but should simply modify the module that is working on. Unfortunately, I think that DebugHelper, as it is designed now, it is not suited for doing this (also here, correct me if I'm wrong, I will be happy about that:)).
The point is that, the `Annotator`, which is the class doing the work of attaching the Metadata to the IR, works by knowing the line at which a certain LLVM Instruction will be in the output `.ll` file (so that the debugger will know at which line to look for showing the corresponding IR instruction when stepping). In fact, in the `generateDebugInfo` method you can see that we serialize the module twice, one just to recover the `DILocation`, and the next to actually serialize the module containing the debug infos (the parameter passed to the annotator handle this behavior):
```
    raw_null_ostream NullStream;
    TheModule->print(NullStream, annotator(true /* DebugInfo */));

    std::ofstream Output(DebugPath);
    raw_os_ostream Stream(Output);
    TheModule->print(Stream, annotator(false));
```
Therefore, I do not see how we can adapt this class to not work on streams but to work on the IR itself, taking into consideration that if the module is later modified in the pipeline, I expected the debug infos to go out-of-sync very fast.

Reply:
Yeah, you're probably right. It wouldn't make sense to have such pass scheduled, if not last.